pyproject.toml

```
@@ -198,6 +198,21 @@ keep-runtime-typing = true
 [tool.ruff.lint.pydocstyle]
 convention = "google"
 
+[tool.pyright]
+include = ["lib/charms/operator_libs_linux/v*/*.py", "tests/*/*.py"]
+pythonVersion = "3.8"  # check no python > 3.8 features are used
+pythonPlatform = "All"
+typeCheckingMode = "strict"
```

I think we've kinda regretted using "strict" in Ops, as it can cause more churn than needed. Or I could be remembering wrong. I wonder how many of the `reportFoo = false`s below would go ahead if we used normal rather than strict mode?

---

Interesting. Looking [here](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#diagnostic-settings-defaults) it does seem like there are a lot of rules that we would lose if we dropped to "basic" type checking. It would indeed remove the need for the manually disabling the rules that we disable currently though.

I also [learned](https://github.com/microsoft/pyright/discussions/4715) that if we want to do gradual type annotations for this repository, it might be worth setting things to "basic" by default, and manually marking individual libraries (like snap) as being compliant with "strict" type checking.

---

I'm leaning away from the "basic" by default approach for this PR, as I'm not clear how to configure rules per level without having to use comments to configure it in each file -- e.g. make unnecessary type ignore an error in "strict" but not in "basic". PRs adding coverage for additional libs can always make this change